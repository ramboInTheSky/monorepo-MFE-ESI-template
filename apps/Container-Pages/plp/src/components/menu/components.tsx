import styled from "styled-components"

export const StyledMenuContainer = styled.div`
    flex: 1;
`
