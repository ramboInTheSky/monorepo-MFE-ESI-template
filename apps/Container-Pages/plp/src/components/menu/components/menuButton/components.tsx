import styled from "styled-components"
import {Button} from "@mui/material"

export const StyledMenuButton = styled(Button)`
    width: 100%;
    height: 2.75rem;
    padding: 0;
    text-transform: uppercase;
    background: none;
    background-color: transparent;
    &:hover {
        background-color: transparent;
    }
    &&:focus {
        border: none;
    }
    color: ${props => props.theme.colours.text.default};
`
