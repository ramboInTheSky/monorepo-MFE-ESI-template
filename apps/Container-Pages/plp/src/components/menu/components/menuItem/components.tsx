import styled from "styled-components"
import {MenuItem, ListItemText, ListItemIcon} from "@mui/material"

import withStyles from "@mui/styles/withStyles"

export const StyledMenuItem = styled(MenuItem)`
    padding: 0.8125rem 0.9375rem 0.8125rem 0.625rem;
`

const listItemTextStyles = {
    root: {
        margin: 0,
        "&$inset": {
            paddingLeft: "1.875rem",
        },
    },
    inset: {},
}
export const StyledListItemText = withStyles(listItemTextStyles)(ListItemText)

export const StyledListItemIcon = styled(ListItemIcon)`
    min-width: 1.875rem;
`
