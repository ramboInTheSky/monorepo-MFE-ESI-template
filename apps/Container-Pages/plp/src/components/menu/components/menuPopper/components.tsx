import styled from "styled-components"
import {Popper, MenuItem, MenuList} from "@mui/material"

export const StyledPopper = styled(Popper)`
    z-index: 1;

    &[data-popper-placement*="bottom"] .arrow {
        width: 0;
        height: 0;
        border-bottom: 0.625rem solid ${props => props.theme.colours.text.default};
        border-left: 0.625rem solid transparent;
        border-right: 0.625rem solid transparent;
        margin-top: -0.625rem;
    }
`

export const StyledMenuItem = styled(MenuItem)`
    padding: 0.8125rem 0.9375rem 0.8125rem 0.625rem;
`

export const Arrow = styled.span`
    position: absolute;
    font-size: 0.4375rem;
    width: 3rem;
    height: 3rem;
`

export const StyledMenuList = styled(MenuList)`
    border-top: 0.25rem solid ${props => props.theme.colours.text.default};
    margin-top: 0.875rem;
    padding: 0;
`
