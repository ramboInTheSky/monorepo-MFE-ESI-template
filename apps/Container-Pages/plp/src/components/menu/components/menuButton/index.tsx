import React from "react"
import {formatTextTestIds} from "@monorepo/utils"
import { TextModel } from "models/Text"
import {StyledMenuButton} from "./components"

interface MenuButtonProps {
    setOpen: (value: React.SetStateAction<boolean>) => void
    anchorRef: React.RefObject<HTMLButtonElement>
    buttonText: string
    open: boolean
    text: TextModel
}
const MenuButton = ({setOpen, anchorRef, buttonText, open, text}: MenuButtonProps) => {
    const handleToggle = () => {
        setOpen(prevOpen => !prevOpen)
    }
    const prevOpen = React.useRef(open)
    React.useEffect(() => {
        if (prevOpen.current === true && open === false) {
            anchorRef.current!.focus()
        }

        prevOpen.current = open
    }, [open, anchorRef])

    return (
        <StyledMenuButton
            aria-controls={open ? "menu-list-grow" : undefined}
            aria-haspopup="true"
            aria-describedby={open ? "styled-options-menu" : undefined}
            onClick={handleToggle}
            disableRipple
            disableElevation
            variant="text"
            ref={anchorRef}
            data-testid={formatTextTestIds(`plp-menu-button-${buttonText}`)}
        >
            {open ? text.labels.close : buttonText}
        </StyledMenuButton>
    )
}
export default MenuButton
