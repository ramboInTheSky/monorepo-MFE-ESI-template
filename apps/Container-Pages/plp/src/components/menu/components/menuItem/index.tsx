import React, {useCallback} from "react"
import {StyledMenuItem, StyledListItemIcon, StyledListItemText} from "./components"
import env from "../../../../config/env"

interface MenuItemProps {
    name: string
    value: string
    isSelected: boolean
    handleClose: (event, value?: string | undefined) => any
}

const MenuItem = ({name, value, isSelected, handleClose}: MenuItemProps) => {
    const onClick = useCallback(
        event => {
            handleClose(event, value)
        },
        [value, handleClose],
    )

    return (
        <StyledMenuItem
            onClick={onClick}
            className={isSelected ? "selected" : ""}
            selected={isSelected}
            autoFocus={isSelected}
        >
            {isSelected && (
                <StyledListItemIcon>
                    <img src={`${env.REACT_APP_BLOB_STORAGE_PATH}/icons/shared/sort-check.svg`} alt="check icon" />
                </StyledListItemIcon>
            )}
            <StyledListItemText primary={name} primaryTypographyProps={{variant: "h5"}} inset={!isSelected} />
        </StyledMenuItem>
    )
}

export default MenuItem
