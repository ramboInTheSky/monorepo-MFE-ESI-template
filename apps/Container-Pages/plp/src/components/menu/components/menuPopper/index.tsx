import {formatTextTestIds} from "@monorepo/utils"
import React, {useCallback} from "react"
import {Grow, Paper, ClickAwayListener} from "@mui/material"
import {StyledPopper, Arrow, StyledMenuList} from "./components"
import {Sorting} from "../../../../models/Sorting"
import MenuItem from "../menuItem"

interface MenuPopperProps {
    buttonText: string
    open: boolean
    anchorRef: React.RefObject<HTMLButtonElement>
    setOpen: (value: React.SetStateAction<boolean>) => void
    options: Sorting
    onSelect: (value: string) => void
}

const MenuPopper = ({buttonText, open, anchorRef, setOpen, options, onSelect}: MenuPopperProps) => {
    const [arrowRef, setArrowRef] = React.useState<null | HTMLElement>(null)

    const handleClose = useCallback(
        (event: MouseEvent | TouchEvent, value?: string) => {
            if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
                return
            }
            setOpen(false)
            if (value && value !== options?.selected) {
                onSelect(value)
            }
        },
        [anchorRef, options, setOpen, onSelect],
    )

    const handleListKeyDown = useCallback(
        (event: React.KeyboardEvent) => {
            if (event.key === "Tab") {
                event.preventDefault()
                setOpen(false)
            }
        },
        [setOpen],
    )

    const menuBody = useCallback(
        ({TransitionProps}) => {
            return (
                <Grow {...TransitionProps} style={{transformOrigin: "center top"}}>
                    <Paper elevation={3}>
                        <Arrow className="arrow" ref={setArrowRef} />
                        <ClickAwayListener onClickAway={handleClose}>
                            <StyledMenuList
                                autoFocusItem={open}
                                id="menu-list-grow"
                                data-testid={formatTextTestIds("menu-list-grow")}
                                onKeyDown={handleListKeyDown}
                            >
                                {options.options.map(so => (
                                    <MenuItem
                                        key={so.value}
                                        name={so.name}
                                        value={so.value}
                                        isSelected={so.value === options?.selected}
                                        handleClose={handleClose}
                                    />
                                ))}
                            </StyledMenuList>
                        </ClickAwayListener>
                    </Paper>
                </Grow>
            )
        },
        [handleClose, handleListKeyDown, open, options],
    )

    return (
        <StyledPopper
            id="styled-options-menu"
            data-testid={formatTextTestIds(`plp-menu-options-${buttonText}`)}
            open={open}
            anchorEl={anchorRef.current}
            role={undefined}
            transition
            disablePortal
            modifiers={[
                {
                    name: "arrow",
                    enabled: true,
                    options: {
                        element: arrowRef,
                    },
                },
                {
                    name: "preventOverflow",
                    options: {
                        padding: 0,
                    },
                },
                {
                    name: "flip",
                    enabled: false,
                },
            ]}
        >
            {menuBody}
        </StyledPopper>
    )
}

export default MenuPopper
