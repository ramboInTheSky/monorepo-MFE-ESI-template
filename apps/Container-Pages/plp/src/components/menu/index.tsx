import React, {useCallback} from "react"
import { TextModel } from "models/Text"
import {StyledMenuContainer} from "./components"

import MenuButton from "./components/menuButton"
import MenuPopper from "./components/menuPopper"
import {Sorting} from "../../models/Sorting"
import TrackSortOpen from "../../events/trackEvent/events/trackSortOpen"

interface MenuProps {
    options?: Sorting
    buttonText: string
    onSelect: (value: string) => void
    text: TextModel
}
// This seems to be the simpliest implementation of a material UI Menu/Tooltip with an arrow.
// See https://github.com/mui-org/material-ui/blob/master/docs/src/pages/components/popper/ScrollPlayground.js

export const Menu = (props: MenuProps) => {
    const {options, onSelect, buttonText, text} = props
    const anchorRef = React.useRef<HTMLButtonElement>(null)

    const [open, setOpen] = React.useState(false)

    const setOpenHandler = useCallback((isOpen) => {
        setOpen(isOpen)
        if (isOpen) TrackSortOpen()
    }, [setOpen])

    return (
        <StyledMenuContainer>
            <MenuButton setOpen={setOpenHandler} anchorRef={anchorRef} buttonText={buttonText} open={open} text={text}/>
            <MenuPopper
                anchorRef={anchorRef}
                buttonText={buttonText}
                open={open}
                setOpen={setOpenHandler}
                options={options!}
                onSelect={onSelect}
            />
        </StyledMenuContainer>
    )
}
export default Menu
