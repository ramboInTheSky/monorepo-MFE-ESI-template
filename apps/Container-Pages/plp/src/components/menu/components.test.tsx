import React from "react"
import {render} from "@testing-library/react"
import {StyledMenuContainer} from "./components"

describe("Snapshots - Sort", () => {
    test("StyledMenuContainer", () => {
        const {asFragment} = render(<StyledMenuContainer>Test</StyledMenuContainer>)
        expect(asFragment()).toMatchSnapshot()
    })
})
