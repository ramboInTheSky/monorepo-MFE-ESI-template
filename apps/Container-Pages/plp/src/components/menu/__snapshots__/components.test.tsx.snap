// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots - Sort StyledMenuContainer 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
    class="c0"
  >
    Test
  </div>
</DocumentFragment>
`;
